// sd-core/core.js
module.exports = {
    sd: new SimpleDash()
};

function SimpleDashWidget (simpleDash, index, config) {
    this.simpleDash = simpleDash;
    this.index = index;
    this.config = config;
    this.dataProvider = config["widget"]["dataprovider"];
    
    if (!this.config["widget"]["type"]) {
        this.config["widget"]["type"] = "chart";
    }
    this.type = this.config["widget"]["type"];
    
    this.data = null;
    
    this.getPlaceholder = function () {
        var placeholder = [
                "<div class='widgetHolder resizable' id='chart",
                this.index,
                "' style='max-height:",
                this.config["widget"]["height"]
                ,"px; max-width:",
                this.config["widget"]["width"]
                ,"px'></div>"].join("");
        return placeholder;
    }
    
    this.fetchData = function () {
        var dp = this.simpleDash.getDataProvider(this.dataProvider);
        if (!dp) {
            throw new Error ("Unknown data provider " + this.dataProvider + " for widget " + this.index);
        }
        
        dp.fetchData(this.index, this.config, this, this.dataAvailable);
    }
    
    this.dataAvailable = function (data) {
        this.data = data;
        this.populateWidget();
    }
    
    this.populateWidget = function () {
        var display = this.simpleDash.getWidgetType(this.type);
        display.populate(window.$("#workspace #chart" + this.index), this.config, this.data);
    }
}

function SimpleDash() {
    this.charts = [];

    this.configObj = {};
    this.chartConfig = [];
    this.chartWidgets = [];

    this.go = function () {
        // 1. read configuration file
        this.readConfiguration();
        // 2. prepare all chart widget
        var placeholders = this.prepareChartWidgets();
        // 3. draw
        this.drawChartWidgets(placeholders);
        // 4. populate with data
        this.populateChartWidgets();
        // 5. apply general layout
        this.applyLayout();
    };
    
    /**
     * Read configuration file using Node.js API
     */
    this.readConfiguration = function () {
        //var app = require("nw.gui").App;
        // XXX: hack for nested modules
        var app = window.nwDispatcher.requireNwGui().App;
        
        
        var config = "../simpledash-config.json";
        
        if (app.argv.length > 0) {
            config = app.argv[0];
        }
        
        var fs = require("fs");
        var configFile = fs.readFileSync("../simpledash-config.json", {encoding: "utf8"});
        
        this.configObj = JSON.parse(configFile);
        for (var i = 0; i < this.configObj.charts.length; i++) {
            this.charts.push(new SimpleDashWidget(this, i, this.configObj.charts[i]));
        }
        this.chartConfig = this.configObj.charts;
    };
    
    /**
     * Prepare placeholders for each widget.
     */
    this.prepareChartWidgets = function () {
        var chartWidgets = [];
        for (var i = 0; i < this.charts.length; i++) {
            // 1. create placeholder
            chartWidgets.push(this.charts[i].getPlaceholder());
        }
        return chartWidgets;
    };
    
    /**
     * Draw everything
     */
    this.drawChartWidgets = function (placeholders) {
        var template = require("./layouts")[this.configObj["layout"]].template;
        window.$("#workspace")[0].innerHTML = window.Handlebars.compile(template)({"charts": placeholders});
    };
    
    /**
     * Initialize each chart
     */
    this.populateChartWidgets = function () {
        for (var i = 0; i < this.charts.length; i++) {
            var widget = this.charts[i];
            widget.fetchData();
        }
    }
    
    /**
     * Apply general layout
     */
    this.applyLayout = function () {
        require("./layouts")[this.configObj["layout"]].initFunction();

        if (this.configObj["resizable"]) {
            $(".resizable").resizable();
        }
    }
    
    this.getDataProvider = function (name) {
        return require("sddp-" + name).dataprovider();
    }
    
    this.clearCharts = function () {
        for (var i = 0; i < this.chartConfig.length; i++) {
            this.widgetTypes[this.chartConfig[i]["widget"]["type"]].clear($("#chart" + i), this.chartConfig[i]);
        }
    };
    
    this.getWidgetType = function (name) {
        return require("./widgets")[name]();
    }
}
