// sd-core/layouts.js
module.exports = {
    "wookmark": new SimpleDashLayout(
        "<ul class='wookmark'>{{#each charts}}<li>{{{this}}}</li>{{/each}}</ul>",
        function () {
            window.$("ul.wookmark li").wookmark({autoResize: true, itemWidth: 200});
        }),
    "2x2": new SimpleDashLayout(
        "<table width='100%' height='100%' id='grid2x2'><tr id='row0'>{{#each charts}}<td id='cell{{@index}}'>{{{this}}}</td>{{/each}}</tr><tr id='row1'></tr></table>",
        function () {
            // re-shuffle cells to their proper place
            window.$("#grid2x2 #cell2").appendTo("#grid2x2 #row1");
            window.$("#grid2x2 #cell3").appendTo("#grid2x2 #row1");
        }),
    "3x2": new SimpleDashLayout(
        "<table width='100%' height='100%' id='grid3x2'><tr id='row0'>{{#each charts}}<td id='cell{{@index}}'>{{{this}}}</td>{{/each}}</tr><tr id='row1'></tr></table>",
        function () {
            // re-shuffle cells to their proper place
            window.$("#grid3x2 #cell3").appendTo("#grid3x2 #row1");
            window.$("#grid3x2 #cell4").appendTo("#grid3x2 #row1");
            window.$("#grid3x2 #cell5").appendTo("#grid3x2 #row1");
        }),
    "masonry": new SimpleDashLayout(
        "<div id='masonry'>{{#each charts}}<div class='item'>{{{this}}}</div>{{/each}}</div>",
        function () {
            window.$("div#masonry").masonry({
                itemSelector: ".item",
                columnWidth: 400, // XXX: make configurable
                gutter: 10
            });
        }),
    "gridster": new SimpleDashLayout(
        "<div id='gridster'><ul>{{#each charts}}<li>{{{this}}}</li>{{/each}}</ul></div>",
        function () {
            window.$("div#gridster ul").gridster({
                widget_margins: [10, 10],
                widget_base_dimensions: [400, 400]
            });
        })
};

function SimpleDashLayout (template, initFunction) {
    this.template = template;
    this.initFunction = initFunction;
}
